src/libs/actions/App.js

```diff
@@ -204,14 +203,79 @@ function reconnectApp(updateIDFrom = 0, updateIDTo = 0) {
+        {
+            updateIDFrom,
+            updateIDTo,
+        },
+        getOnyxDataForOpenOrReconnect(),
+
+        // Set this to true so that the request will be prioritized at the front of the sequential queue
+        true,
+    );
+}
+
+// The next 40ish lines of code are used for detecting when there is a gap of OnyxUpdates between what was last applied to the client and the updates the server has.
+// When a gap is detected, the missing updates are fetched from the API.
+
+// These key needs to be separate from ONYXKEYS.ONYX_UPDATES so that it can be updated without triggering the callback when the server IDs are updated
+let lastUpdateIDAppliedToClient = 0;
+Onyx.connect({
+    key: ONYXKEYS.ONYX_UPDATES_LAST_UPDATE_ID_APPLIED_TO_CLIENT,
+    callback: (val) => (lastUpdateIDAppliedToClient = val),
+});
+
+// This object is stored in Onyx and the callback is triggered anytime new update IDs are received either from Pusher or from HTTPs (it was the only way to keep the code DRY and to prevent circular dependencies)
+const onyxUpdates = {
+    lastUpdateIDFromServer: 0,
+    previousUpdateIDFromServer: 0,
+};
+Onyx.connect({
+    key: ONYXKEYS.ONYX_UPDATES,
+    callback: (val) => {
+        if (!val) {
+            return;
+        }
+
+        const {lastUpdateIDFromServer, previousUpdateIDFromServer} = val;
+        console.debug('[OnyxUpdates] Received lastUpdateID from server', lastUpdateIDFromServer);
+        console.debug('[OnyxUpdates] Received previousUpdateID from server', previousUpdateIDFromServer);
+        console.debug('[OnyxUpdates] Last update ID applied to the client', lastUpdateIDAppliedToClient);
+
+        // If the previous update from the server does not match the last update the client got, then the client is missing some updates.
+        // getMissingOnyxUpdates will fetch updates starting from the last update this client got and going to the last update the server sent.
+        if (lastUpdateIDAppliedToClient && previousUpdateIDFromServer && lastUpdateIDAppliedToClient < previousUpdateIDFromServer) {
+            console.debug('[OnyxUpdates] Gap detected in update IDs so fetching incremental updates');
+            Log.info('Gap detected in update IDs from Pusher so fetching incremental updates', true, {
+                lastUpdateIDFromServer,
+                previousUpdateIDFromServer,
+                lastUpdateIDAppliedToClient,
+            });
+            getMissingOnyxUpdates(lastUpdateIDAppliedToClient, lastUpdateIDFromServer);
+        }
+
+        // Update the local values to be the same as the values stored in Onyx
+        onyxUpdates.lastUpdateIDFromServer = lastUpdateIDFromServer || 0;
+        onyxUpdates.previousUpdateIDFromServer = previousUpdateIDFromServer || 0;
+
+        // Update this value so that it matches what was just received from the server
+        Onyx.merge(ONYXKEYS.ONYX_UPDATES_LAST_UPDATE_ID_APPLIED_TO_CLIENT, lastUpdateIDFromServer || 0);
+    },
```

Comment:
From your PR I was able to see something I wasn't seeing before: We also send GetOnyxUpdate responses through Pusher, so we're updating the current client and sending those updates to other clients. I think that's fine since if we're sending it to all clients, the changes the clients are updated are bigger.

What that does is:

- Client A and B are using account 1
- Client A has OnyxUpdatesLastUpdateIDAppliedToClient with 51
- Client B as OnyxUpdatesLastUpdateIDAppliedToClient with 47
- Account 2 sends a message to Account 1 which will be updateID 55
- Client B for some reason doesn't receive it
- Client A receives, and calls `GetOnyxUpdates` with `from: 51` and `to: 55`
- Client B, that has it's connection working now, receives the `GetOnyxUpdates` from Pusher, sees that the previousUpdateID in that update is bigger than 47, and calls GetOnyxUpdates with values `from: 47` and `to: 51`, and gets up to date.

This works OK!

Now we have an edge case that happened when I was testing.

- Client A and B are using account 1
- Client B as OnyxUpdatesLastUpdateIDAppliedToClient with 47
- Account 2 sends a message to Account 1 which will be updateID 55
- Client A receives it instantly and now has OnyxUpdatesLastUpdateIDAppliedToClient with 55. Then it does some actions, and that number rises to 60
- Client B for some reason doesn't receive those
- Client B connects and receives update 60. It notices it didn't have 47-59, and does a `GetOnyxUpdates` with those
- Client A receives those updates since we're testing. It won't apply those, since the ID is less than the one it currently has BUT it will save `OnyxUpdatesLastUpdateIDAppliedToClient` with lastUpdateID 58 that was sent in the pusher update body.
- So next time it tries to do anything, it will call `GetMissingOnyxUpdates` again


I think we solve that by adding this to the if statement above.

Reply:
OK, to make sure I understand your proposed change clearly, you are suggesting to only update `OnyxUpdatesLastUpdateIDAppliedToClient` if the `lastUpdateIDFromServer` is larger. Right? That way the value will never decrease, only increase.

Reply:
Yes, that's correct!